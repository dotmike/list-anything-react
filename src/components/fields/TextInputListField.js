import React from "react";
import PropTypes from "prop-types";
import { Form } from "semantic-ui-react";

import ListValues from "./ListValues";
import EditableListInput from "./EditableListInput";

const styles = {
  showFieldDescription: {
    marginTop: "0px",
    maxHeight: "100px",
    overflowY: "auto"
  },
  default: { maxHeight: "100px", overflowY: "auto", minWidth: "100%" }
};

class TextInputListField extends React.Component {
  constructor(props) {
    super(props);
    this.values = props.value ? props.value.split("\n") : [];
  }

  shouldComponentUpdate(nextProps) {
    return (
      this.props.editable !== nextProps.editable ||
      this.props.showFieldDescription !== nextProps.showFieldDescription ||
      this.props.error !== nextProps.error
    );
  }

  onChange = e => {
    this.values = e.target.values;
    this.props.onChange(e);
  };

  render() {
    const { editable, showFieldDescription, error, field } = this.props;
    const { values } = this;

    if (editable)
      return (
        <EditableListInput
          values={values}
          type="text"
          error={error}
          field={field}
          onChange={this.onChange}
        />
      );
    if (showFieldDescription)
      return (
        <Form.Field>
          <b>{field.description}:</b>
          <ListValues values={values} style={styles.showFieldDescription} />
        </Form.Field>
      );
    return <ListValues values={values} style={styles.default} />;
  }
}

TextInputListField.propTypes = {
  // ownProps
  value: PropTypes.string.isRequired,
  editable: PropTypes.bool,
  showFieldDescription: PropTypes.bool,
  error: PropTypes.string,
  field: PropTypes.shape({
    _id: PropTypes.string,
    description: PropTypes.string
  }),
  onChange: PropTypes.func
};

TextInputListField.defaultProps = {
  // ownProps
  editable: false,
  showFieldDescription: true,
  error: "",
  field: {},
  onChange: () => {}
};

export default TextInputListField;
