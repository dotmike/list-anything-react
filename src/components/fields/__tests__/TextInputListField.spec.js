import TextInputListField from "../TextInputListField";

const testData = {
  value: "testando...",
  field: { _id: "1", description: "Testando:" },
  newValue: "test"
};

const setup = (propOverrides = {}) => {
  const props = {
    value: testData.value,
    error: "",
    editable: true,
    showFieldDescription: true,
    field: testData.field,
    onChange: jest.fn(),
    ...propOverrides
  };

  return {
    props,
    wrapperShallow: wrapperShallow(TextInputListField, props)
  };
};

describe("TextInputListField", () => {
  describe("when `editable` is true", () => {
    it("renders correctly", () => {
      const { wrapperShallow: wrapper } = setup({
        editable: true
      });
      expect(wrapper()).toMatchSnapshot();
    });

    it("renders correctly when `error` is passed", () => {
      const { wrapperShallow: wrapper } = setup({
        editable: true,
        error: "Something went wrong..."
      });
      expect(wrapper()).toMatchSnapshot();
    });
  });

  describe("when `editable` is false", () => {
    it("renders correctly when `showFieldDescription` is true", () => {
      const { wrapperShallow: wrapper } = setup({
        editable: false,
        showFieldDescription: true
      });
      expect(wrapper()).toMatchSnapshot();
    });

    it("renders correctly when `showFieldDescription` is false", () => {
      const { wrapperShallow: wrapper } = setup({
        editable: false,
        showFieldDescription: false
      });
      expect(wrapper()).toMatchSnapshot();
    });
  });

  it("shoudn't update when `value` prop is changed", () => {
    const spy = jest.spyOn(TextInputListField.prototype, "render");
    const { wrapperShallow: wrapper } = setup();

    wrapper().setProps({ value: testData.newValue });
    expect(spy).toHaveBeenCalledTimes(1);

    wrapper().setProps({ error: "Something went wrong..." });
    expect(spy).toHaveBeenCalledTimes(2);
  });
});
