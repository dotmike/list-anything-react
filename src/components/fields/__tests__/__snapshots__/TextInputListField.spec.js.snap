// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputListField when \`editable\` is false renders correctly when \`showFieldDescription\` is false 1`] = `
<ListField
  onRemove={null}
  renderValue={[Function]}
  style={
    Object {
      "maxHeight": "100px",
      "minWidth": "100%",
      "overflowY": "auto",
    }
  }
  values={
    Array [
      "testando...",
    ]
  }
/>
`;

exports[`TextInputListField when \`editable\` is false renders correctly when \`showFieldDescription\` is true 1`] = `
<FormField>
  <b>
    Testando:
    :
  </b>
  <ListField
    onRemove={null}
    renderValue={[Function]}
    style={
      Object {
        "marginTop": "0px",
        "maxHeight": "100px",
        "overflowY": "auto",
      }
    }
    values={
      Array [
        "testando...",
      ]
    }
  />
</FormField>
`;

exports[`TextInputListField when \`editable\` is true renders correctly 1`] = `
<EditableListInput
  error=""
  field={
    Object {
      "_id": "1",
      "description": "Testando:",
    }
  }
  onChange={[Function]}
  type="text"
  values={
    Array [
      "testando...",
    ]
  }
/>
`;

exports[`TextInputListField when \`editable\` is true renders correctly when \`error\` is passed 1`] = `
<EditableListInput
  error="Something went wrong..."
  field={
    Object {
      "_id": "1",
      "description": "Testando:",
    }
  }
  onChange={[Function]}
  type="text"
  values={
    Array [
      "testando...",
    ]
  }
/>
`;
